nexus/mgs-updates/src/host_phase1_updater.rs

```diff
@@ -3,20 +3,153 @@
+//!      that other MGS-driven updates are done: the currently-active slot (B)
+//!      now contains the new version (X1). However, we must not return
+//!      `UpdateComplete` from precheck yet, so that we can handle Nexus dying
+//!      before we finish each of the next two steps. To handle this, once the
+//!      phase 1 update looks complete, we also query the sled-agent to check
+//!      that its boot disk matches the currently-active slot. A precheck at
+//!      this point will fail with `MismatchedHostOsActiveSlot`, because the
+//!      sled's boot disk (A) doesn't match the active phase 1 slot (B).
+//!
+//!    * We power off the host (i.e., go to PowerState::A2). This does not
+//!      change the state of the sled, except that we can no longer talk to
+//!      sled-agent and therefore no longer have a report of the boot disk.
+//!
+//!      A precheck at this point will fail with `SledAgentInventory`, because
+//!      we won't be able to fetch inventory from sled-agent on a powered-off
+//!      sled.
+//!
+//!    * We power on the host (i.e., go to PowerState::A0). Once it comes back,
+//!      the state of the sled will be:
+//!
+//!      - active phase 1 slot: B
+//!      - boot disk: B
+//!      - slot A contains Y1+Y2
+//!      - slot B contains X1+X2
+//!
+//!      and the upgrade will be complete.
 
 use super::MgsClients;
 use super::SpComponentUpdateError;
 use super::UpdateProgress;
 use super::common_sp_update::SpComponentUpdater;
 use super::common_sp_update::deliver_update;
+use crate::SpComponentUpdateHelperImpl;
+use crate::common_sp_update::PostUpdateError;
+use crate::common_sp_update::PrecheckError;
+use crate::common_sp_update::PrecheckStatus;
+use futures::FutureExt as _;
+use futures::future::BoxFuture;
+use gateway_client::HostPhase1HashError;
 use gateway_client::SpComponent;
 use gateway_client::types::SpComponentFirmwareSlot;
 use gateway_client::types::SpType;
+use nexus_sled_agent_shared::inventory::BootPartitionContents;
+use nexus_types::deployment::PendingMgsUpdate;
+use nexus_types::deployment::PendingMgsUpdateHostPhase1Details;
+use omicron_common::disk::M2Slot;
+use sled_agent_client::Client as SledAgentClient;
 use slog::Logger;
+use slog::debug;
 use slog::info;
+use slog_error_chain::InlineErrorChain;
+use std::time::Duration;
 use tokio::sync::watch;
+use tufaceous_artifact::ArtifactHash;
+use tufaceous_artifact::ArtifactKind;
 use uuid::Uuid;
 
+// Hashing the current phase 1 contents on the SP is an asynchronous operation:
+// we request a hash and then poll until the hashing completes. We have to pick
+// some timeout to give up on polling. In practice we expect this hashing to
+// take a few seconds, so set something very generous here that would indicate
+// something very wrong.
+const PHASE_1_HASHING_TIMEOUT: Duration = Duration::from_secs(60);
+
 type GatewayClientError = gateway_client::Error<gateway_client::types::Error>;
 
 pub struct HostPhase1Updater {
```

Comment:
I guess we're coming up on time to rip out these older updaters, right?  I think I wanted to keep them in previously because they sort of tested the SP behavior.  But once we've implemented the new updaters and their tests that use the SP behavior, I can't think of a reason to keep those (or their tests) around.

Reply:
Yeah, agreed.